src/workerd/io/actor-sqlite-test.c++

```diff
@@ -0,0 +1,440 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "actor-sqlite.h"
+#include <kj/test.h>
+#include <kj/debug.h>
+#include "io-gate.h"
+#include <workerd/util/capnp-mock.h>
+#include <workerd/util/test.h>
+
+namespace workerd {
+namespace {
+
+static constexpr kj::Date oneMs = 1 * kj::MILLISECONDS + kj::UNIX_EPOCH;
+static constexpr kj::Date twoMs = 2 * kj::MILLISECONDS + kj::UNIX_EPOCH;
+
+template <typename T>
+kj::Promise<T> eagerlyReportExceptions(kj::Promise<T> promise, kj::SourceLocation location = {}) {
+  return promise.eagerlyEvaluate([location](kj::Exception&& e) -> T {
+    KJ_LOG_AT(ERROR, location, e);
+    kj::throwFatalException(kj::mv(e));
+  });
+}
+
+// Expect that a synchronous result is returned.
+template <typename T>
+T expectSync(kj::OneOf<T, kj::Promise<T>> result, kj::SourceLocation location = {}) {
+  KJ_SWITCH_ONEOF(result) {
+    KJ_CASE_ONEOF(promise, kj::Promise<T>) {
+      KJ_FAIL_ASSERT_AT(location, "result was unexpectedly asynchronous");
+    }
+    KJ_CASE_ONEOF(value, T) {
+      return kj::mv(value);
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+struct ActorSqliteTestOptions final {
+  bool monitorOutputGate = true;
+};
+
+class ActorSqliteTestHooks;
+
+struct ActorSqliteTest final {
+  kj::EventLoop loop;
+  kj::WaitScope ws;
+
+  OutputGate gate;
+  kj::Own<const kj::Directory> vfsDir;
+  SqliteDatabase::Vfs vfs;
+  SqliteDatabase db;
+
+  kj::Vector<kj::Own<kj::PromiseFulfiller<void>>> commitFulfillers;
+  kj::Vector<kj::String> calls;
+
+  class ActorSqliteTestHooks final: public ActorSqlite::Hooks {
+  public:
+    ActorSqliteTestHooks(ActorSqliteTest& parent): parent(parent) {}
+
+    kj::Promise<kj::Maybe<kj::Date>> getAlarm() override {
+      parent.calls.add(kj::str("getAlarm"));
+      return kj::Maybe<kj::Date>(kj::none);
+    }
+
+    kj::Promise<void> setAlarm(kj::Maybe<kj::Date> newAlarmTime) override {
+      auto time = newAlarmTime.map([](auto& t) { return kj::str(t); }).orDefault(kj::str("none"));
+      parent.calls.add(kj::str("setAlarm(", time, ")"));
```

Comment:
minor nit...

```suggestion
      auto time = newAlarmTime.map([](auto& t) {
        return kj::str("setAlarm(", t, ")");
      }).orDefault(kj::str("setAlarm(\"none\")"));
      parent.calls.add(kj::mv(time));
```

Should save an extraneous copy and allocation. Obviously super minor tho.